pkg/cmd/roachtest/quit.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/contextutil"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/errors"
+)
+
+// runQuitTransfersLeases performs rolling restarts on a
+// 3-node cluster and ascertains that each node shutting down
+// transfers all its leases reliably to other nodes prior to
+// terminating.
+func runQuitTransfersLeases(
+	ctx context.Context,
+	t *test,
+	c *cluster,
+	methodName string,
+	method func(ctx context.Context, t *test, c *cluster, nodeID int),
+) {
+	args := startArgs("--env=COCKROACH_SCAN_MAX_IDLE_TIME=5ms")
+	c.Put(ctx, cockroach, "./cockroach")
+	c.Start(ctx, t, args)
+
+	// rwt runs a command with a 1-minute timeout.
+	rwt := func(fn func(ctx context.Context)) {
+		if err := contextutil.RunWithTimeout(ctx, "do", time.Minute, func(ctx context.Context) error {
+			fn(ctx)
+			return nil
+		}); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Now start the remainder of the nodes, then wait for
+	// up-replication.
+	func() {
```

Comment:
Could you make a

```
type quitTest struct {
c *cluster
db *gosql.DB // opening one takes 1s...
}
```

and hang these various closures off of `*quitTest` (base signature `(context.Context, *test)`)? I just gave acceptance/version-upgrade that treatment and found it made it much more pleasant to read these tests as you get to put the "meat" of the test early in the file and put all the helpers below.